<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<html><head><title>Python: module tab1ToRef2</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8">
</head><body bgcolor="#f0f0f8">

<table width="100%" cellspacing=0 cellpadding=2 border=0 summary="heading">
<tr bgcolor="#7799ee">
<td valign=bottom>&nbsp;<br>
<font color="#ffffff" face="helvetica, arial">&nbsp;<br><big><big><strong>tab1ToRef2</strong></big></big></font></td
><td align=right valign=bottom
><font color="#ffffff" face="helvetica, arial"><a href=".">index</a><br><a href="file:/work/cbellot/Stage_GenomeHarvest/tab1ToRef2.py">/work/cbellot/Stage_GenomeHarvest/tab1ToRef2.py</a></font></td></tr></table>
    <p></p>
<p>
<table width="100%" cellspacing=0 cellpadding=2 border=0 summary="section">
<tr bgcolor="#aa55cc">
<td colspan=3 valign=bottom>&nbsp;<br>
<font color="#ffffff" face="helvetica, arial"><big><strong>Modules</strong></big></font></td></tr>
    
<tr><td bgcolor="#aa55cc"><tt>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</tt></td><td>&nbsp;</td>
<td width="100%"><table width="100%" summary="list"><tr><td width="25%" valign=top><a href="Bio.SeqIO.html">Bio.SeqIO</a><br>
<a href="argparse.html">argparse</a><br>
</td><td width="25%" valign=top><a href="os.html">os</a><br>
<a href="pydoc.html">pydoc</a><br>
</td><td width="25%" valign=top><a href="re.html">re</a><br>
<a href="sys.html">sys</a><br>
</td><td width="25%" valign=top><a href="tempfile.html">tempfile</a><br>
<a href="warnings.html">warnings</a><br>
</td></tr></table></td></tr></table><p>
<table width="100%" cellspacing=0 cellpadding=2 border=0 summary="section">
<tr bgcolor="#eeaa77">
<td colspan=3 valign=bottom>&nbsp;<br>
<font color="#ffffff" face="helvetica, arial"><big><strong>Functions</strong></big></font></td></tr>
    
<tr><td bgcolor="#eeaa77"><tt>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</tt></td><td>&nbsp;</td>
<td width="100%"><dl><dt><a name="-align"><strong>align</strong></a>(tabOp)</dt><dd><tt>This&nbsp;function&nbsp;make&nbsp;a&nbsp;system&nbsp;call&nbsp;to&nbsp;bwa&nbsp;mem&nbsp;to&nbsp;align<br>
fasta2&nbsp;file&nbsp;with&nbsp;selected&nbsp;sequence&nbsp;of&nbsp;fasta1.&nbsp;The<br>
output&nbsp;file&nbsp;is&nbsp;in&nbsp;sam&nbsp;format.<br>
If&nbsp;argument&nbsp;--notempfile&nbsp;is&nbsp;enable&nbsp;it&nbsp;create&nbsp;the<br>
file&nbsp;"aln_out_[ext].sam".</tt></dd></dl>
 <dl><dt><a name="-checkDependency"><strong>checkDependency</strong></a>()</dt><dd><tt>Check&nbsp;if&nbsp;all&nbsp;third-party&nbsp;program&nbsp;are&nbsp;installed&nbsp;by<br>
making&nbsp;a&nbsp;system&nbsp;call&nbsp;with&nbsp;"module&nbsp;list"&nbsp;then<br>
comparing&nbsp;with&nbsp;a&nbsp;required&nbsp;module&nbsp;list.&nbsp;If&nbsp;at&nbsp;least<br>
one&nbsp;thrird-party&nbsp;program&nbsp;isn't&nbsp;installed,&nbsp;it&nbsp;quit.<br>
Currently,&nbsp;this&nbsp;function&nbsp;check&nbsp;for&nbsp;environment<br>
modules&nbsp;but&nbsp;it&nbsp;need&nbsp;further&nbsp;test&nbsp;or&nbsp;a&nbsp;more&nbsp;generic<br>
test.</tt></dd></dl>
 <dl><dt><a name="-cutGff"><strong>cutGff</strong></a>()</dt><dd><tt>If&nbsp;the&nbsp;argument&nbsp;--typeA&nbsp;is&nbsp;specified,&nbsp;we&nbsp;need&nbsp;to&nbsp;cut<br>
the&nbsp;gff&nbsp;in&nbsp;order&nbsp;to&nbsp;keep&nbsp;only&nbsp;specified&nbsp;type.&nbsp;We&nbsp;use<br>
lower()&nbsp;function&nbsp;on&nbsp;each&nbsp;type&nbsp;in&nbsp;order&nbsp;to&nbsp;be&nbsp;case<br>
insensitive.</tt></dd></dl>
 <dl><dt><a name="-getFasta"><strong>getFasta</strong></a>(tabOP)</dt><dd><tt>This&nbsp;function&nbsp;just&nbsp;extract&nbsp;the&nbsp;sequence&nbsp;defined&nbsp;in<br>
the&nbsp;tabbed&nbsp;file&nbsp;create&nbsp;by&nbsp;<a href="#-getFlank">getFlank</a>().&nbsp;The&nbsp;output&nbsp;is<br>
a&nbsp;fasta&nbsp;file.&nbsp;If&nbsp;--notempfile&nbsp;is&nbsp;enable,&nbsp;it&nbsp;create&nbsp;the<br>
file&nbsp;[fasta1_name]_selected_[ext].fasta&nbsp;in&nbsp;the<br>
directory&nbsp;./result.</tt></dd></dl>
 <dl><dt><a name="-getFlank"><strong>getFlank</strong></a>()</dt><dd><tt>Change&nbsp;the&nbsp;position&nbsp;in&nbsp;tabbed&nbsp;file&nbsp;in&nbsp;order&nbsp;to<br>
include&nbsp;flank&nbsp;region.&nbsp;For&nbsp;vcf&nbsp;file,&nbsp;we&nbsp;need&nbsp;to<br>
generate&nbsp;a&nbsp;new&nbsp;temporary&nbsp;tabbed&nbsp;file&nbsp;which&nbsp;contain<br>
the&nbsp;ID,&nbsp;start&nbsp;position&nbsp;and&nbsp;stop&nbsp;position.&nbsp;If&nbsp;the<br>
argument&nbsp;--notempfile&nbsp;is&nbsp;passed,&nbsp;it&nbsp;create&nbsp;the&nbsp;file<br>
[tabbed_file_name]_out.[ext]&nbsp;in&nbsp;the&nbsp;directory<br>
./result.</tt></dd></dl>
 <dl><dt><a name="-getPosCds"><strong>getPosCds</strong></a>(tab)</dt><dd><tt>This&nbsp;function&nbsp;extract&nbsp;position&nbsp;of&nbsp;mRNA&nbsp;(by&nbsp;defaut,<br>
else&nbsp;it's&nbsp;position&nbsp;of&nbsp;gene)&nbsp;and&nbsp;the&nbsp;relative<br>
position&nbsp;of&nbsp;CDS&nbsp;within.<br>
It&nbsp;take&nbsp;a&nbsp;file&nbsp;in&nbsp;gff&nbsp;format&nbsp;and&nbsp;return&nbsp;a<br>
dictionary&nbsp;with&nbsp;tuple&nbsp;of&nbsp;mRNA&nbsp;(or&nbsp;gene)&nbsp;in&nbsp;keys&nbsp;and<br>
list&nbsp;of&nbsp;list&nbsp;of&nbsp;positon&nbsp;of&nbsp;CDS&nbsp;in&nbsp;value.</tt></dd></dl>
 <dl><dt><a name="-index"><strong>index</strong></a>()</dt><dd><tt>This&nbsp;function&nbsp;generate&nbsp;the&nbsp;index&nbsp;of&nbsp;inputted&nbsp;fasta2<br>
file&nbsp;with&nbsp;a&nbsp;system&nbsp;call&nbsp;to&nbsp;bwa&nbsp;index.</tt></dd></dl>
 <dl><dt><a name="-isComplete"><strong>isComplete</strong></a>(samtotabOut)</dt><dd><tt>This&nbsp;function&nbsp;call&nbsp;the&nbsp;function&nbsp;<a href="#-getPosCds">getPosCds</a>(tab)&nbsp;with<br>
both&nbsp;inputted&nbsp;tabbed&nbsp;file&nbsp;and&nbsp;newly&nbsp;generated&nbsp;tabbed<br>
file&nbsp;and&nbsp;check&nbsp;if&nbsp;CDS&nbsp;in&nbsp;the&nbsp;newly&nbsp;generated&nbsp;file<br>
are&nbsp;in&nbsp;the&nbsp;same&nbsp;position&nbsp;within&nbsp;the&nbsp;mRNA&nbsp;(or&nbsp;gene).</tt></dd></dl>
 <dl><dt><a name="-parseCigar"><strong>parseCigar</strong></a>(sam)</dt><dd><tt>This&nbsp;function&nbsp;parse&nbsp;the&nbsp;CIGAR&nbsp;code&nbsp;in&nbsp;alignment&nbsp;file<br>
in&nbsp;order&nbsp;to&nbsp;get&nbsp;the&nbsp;length&nbsp;of&nbsp;the&nbsp;match.</tt></dd></dl>
 <dl><dt><a name="-parseFa"><strong>parseFa</strong></a>()</dt><dd><tt>Later,&nbsp;in&nbsp;order&nbsp;to&nbsp;get&nbsp;the&nbsp;flank&nbsp;region&nbsp;and&nbsp;not<br>
going&nbsp;over&nbsp;last&nbsp;position&nbsp;of&nbsp;the&nbsp;chromosome,&nbsp;we&nbsp;need<br>
to&nbsp;index&nbsp;the&nbsp;length&nbsp;of&nbsp;them.<br>
This&nbsp;function&nbsp;parse&nbsp;the&nbsp;inputted&nbsp;file&nbsp;fasta1&nbsp;and<br>
write&nbsp;in&nbsp;a&nbsp;tempfile&nbsp;all&nbsp;the&nbsp;chromosome&nbsp;ID&nbsp;and&nbsp;length.</tt></dd></dl>
 <dl><dt><a name="-prefix"><strong>prefix</strong></a>()</dt><dd><tt>In&nbsp;order&nbsp;to&nbsp;link&nbsp;the&nbsp;tabbed&nbsp;file&nbsp;and&nbsp;the&nbsp;fasta<br>
file,&nbsp;we&nbsp;need&nbsp;to&nbsp;get&nbsp;the&nbsp;same&nbsp;chromosome&nbsp;ID.&nbsp;This<br>
function&nbsp;check&nbsp;for&nbsp;this&nbsp;and&nbsp;change&nbsp;ID&nbsp;prefix&nbsp;in&nbsp;the<br>
tabbed&nbsp;file&nbsp;to&nbsp;the&nbsp;same&nbsp;as&nbsp;the&nbsp;ID&nbsp;prefix&nbsp;in&nbsp;the<br>
fasta&nbsp;file.</tt></dd></dl>
 <dl><dt><a name="-samToTab"><strong>samToTab</strong></a>()</dt><dd><tt>In&nbsp;order&nbsp;to&nbsp;get&nbsp;the&nbsp;result&nbsp;of&nbsp;alignment&nbsp;usable&nbsp;for<br>
further&nbsp;&nbsp;use,&nbsp;we&nbsp;create&nbsp;a&nbsp;new&nbsp;tabbed&nbsp;file&nbsp;in&nbsp;the<br>
same&nbsp;format&nbsp;as&nbsp;the&nbsp;inputted&nbsp;tabbed&nbsp;file.&nbsp;This&nbsp;file<br>
is&nbsp;the&nbsp;final&nbsp;output&nbsp;of&nbsp;this&nbsp;program.<br>
First&nbsp;of&nbsp;all,&nbsp;we&nbsp;extract&nbsp;the&nbsp;start&nbsp;and&nbsp;stop&nbsp;position<br>
and&nbsp;remove&nbsp;the&nbsp;flank&nbsp;region&nbsp;to&nbsp;get&nbsp;the&nbsp;initial<br>
annotation&nbsp;length.&nbsp;Then&nbsp;we&nbsp;generate&nbsp;the&nbsp;output&nbsp;file<br>
with&nbsp;these&nbsp;new&nbsp;position&nbsp;and&nbsp;keep&nbsp;all&nbsp;the&nbsp;information<br>
of&nbsp;the&nbsp;original&nbsp;tabbed&nbsp;file.<br>
The&nbsp;name&nbsp;of&nbsp;output&nbsp;file&nbsp;is&nbsp;by&nbsp;defaut<br>
"[fasta2_name]_out.[ext]"&nbsp;or&nbsp;the&nbsp;name&nbsp;that&nbsp;user<br>
specify&nbsp;with&nbsp;argument&nbsp;--output&nbsp;(but&nbsp;still&nbsp;with&nbsp;[ext]<br>
as&nbsp;extension.</tt></dd></dl>
</td></tr></table><p>
<table width="100%" cellspacing=0 cellpadding=2 border=0 summary="section">
<tr bgcolor="#55aa55">
<td colspan=3 valign=bottom>&nbsp;<br>
<font color="#ffffff" face="helvetica, arial"><big><strong>Data</strong></big></font></td></tr>
    
<tr><td bgcolor="#55aa55"><tt>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</tt></td><td>&nbsp;</td>
<td width="100%"><strong>optional</strong> = &lt;argparse._ArgumentGroup object&gt;<br>
<strong>parser</strong> = ArgumentParser(prog='pydoc3', usage=None, descri...tter'&gt;, conflict_handler='error', add_help=False)<br>
<strong>required</strong> = &lt;argparse._ArgumentGroup object&gt;</td></tr></table>
</body></html>